txnsync/bitmask.go

```diff
@@ -0,0 +1,252 @@
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"bytes"
+	"errors"
+)
+
+var errIndexNotFound = errors.New("invalid bitmask: index not found")
+var errInvalidBitmaskType = errors.New("invalid bitmask type")
+
+//msgp:allocbound bitmask maxBitmaskSize
+type bitmask []byte
+
+// assumed to be in mode 0, sets bit at index to 1
+func (b *bitmask) setBit(index int) {
+	byteIndex := index/8 + 1
+	(*b)[byteIndex] |= 1 << (index % 8)
+}
+
+// trimBitmask compresses the bitmask into one of the 4 types:
+// type 0: input bitmask bit pos x b -> output bitmask bit pos x b
+// type 1: input bitmask bit pos x b -> output bitmask bit pos x !b
+// type 2: stores the positions of bits where the bit value is 1
+//         input bitmask first bit 1 at pos A, second bit 1 at pos B, ...
+//         output bitmask stores A, B-A, ...
+// type 3: same as type 2, but stores the positons where the bit is 0
+func (b *bitmask) trimBitmask(entries int) {
+	if *b == nil {
+		return
+	}
+	lastExists := 0
+	lastNotExists := 0
+	numExists := 0
+	for i := 0; i < entries; i++ {
+		byteIndex := i/8 + 1
+		if (*b)[byteIndex]&(1<<(i%8)) != 0 {
+			lastExists = i
+			numExists++
+		} else {
+			lastNotExists = i
+		}
+	}
+	bitmaskType := 0
+	bestSize := bytesNeededBitmask(lastExists)
+	if bestSize > bytesNeededBitmask(lastNotExists) {
+		bitmaskType = 1
+		bestSize = bytesNeededBitmask(lastNotExists)
+	}
+	if bestSize > numExists*2+1 {
+		bitmaskType = 2
+		bestSize = numExists*2 + 1
+	}
+	if bestSize > (entries-numExists)*2+1 {
+		bitmaskType = 3
+		bestSize = (entries-numExists)*2 + 1
+	}
+	switch bitmaskType {
+	case 1:
+		(*b)[0] = 1
+		for i := range *b {
+			if i != 0 {
+				(*b)[i] = 255 - (*b)[i] // invert bits
+			}
+		}
+	case 2:
+		newBitmask := make(bitmask, 1, bestSize)
+		newBitmask[0] = 2
+		last := 0
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if (*b)[byteIndex]&(1<<(i%8)) != 0 {
+				diff := i - last
+				newBitmask = append(newBitmask, byte(diff/256), byte(diff%256))
+				last = i
+			}
+		}
+		*b = newBitmask
+		return
+	case 3:
+		newBitmask := make(bitmask, 1, bestSize)
+		newBitmask[0] = 3
+		last := 0
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if (*b)[byteIndex]&(1<<(i%8)) == 0 {
+				diff := i - last
+				newBitmask = append(newBitmask, byte(diff/256), byte(diff%256))
+				last = i
+			}
+		}
+		*b = newBitmask
+		return
+	default:
+	}
+
+	*b = bytes.TrimRight(*b, "\x00")
```

Comment:
we should just do `*b = (*b)[:bestSize]`

Reply:
I don't see how it will make a difference.

Reply:
It would helps by avoiding scanning the byte array. I think that it *should* be fine - but would be happy to see some testing around that.

Reply:
All we need to do is resize the slice, and we have all the information to do it. There is no need to use a function designed for utf-8.